.. _section-about:

General information
===================

About PARANOiD
--------------
PARANOiD (Pipeline for Automated Read ANalysis Of iCLIP Data) is a fully automated pipeline that analyses reads generated by iCLIP (https://doi.org/10.1016/j.ymeth.2013.10.011) or iCLIP2 (https://doi.org/10.1016/j.ymeth.2019.10.003) experiments. It provides all steps necessary from read preprocessing, over read alignment to peak calling and offers several optional subsequent analyses such as the :ref:`sequence extraction/motif analysis <motif-detection>` or the :ref:`RNA subtype analysis <RNA-subtype-analysis>`. PARANOiD offers users to adapt a large variety of :ref:`options <section-parameters>` and is not restricted to specific reference genomes. While the default aligner is set to Bowtie2 which is only 

About CLIP
----------
CLIP (UV-cross-linking immunoprocipitation) is a method that allows the analysis of RNA protein interaction. It takes advantage of proteins being covalently linked to RNAs in clos proximity when irradiated by UV-light at ~ 254 nm. Via immunoprocipitation a target protein together with all cross-linked RNAs are extracted and after several purification and adapter ligation steps a reverse transcription is performed. The resulting cDNA is amplified via PCR and then sequenced. Most CLIP-methods offer the advantage that they can be performed in vivo as well as in vitro, giving them a broad spectrum of possible applications.  
Since the advent of CLIP a variety of different CLIP variants emerged each with different advantages. 

About iCLIP
-----------
One of these methods is iCLIP (individual-nucleotide resultion CLIP) which takes advantage of the fact that ~80% of all reverse transcrioptions stall at the polypeptide still cross-linked to the RNA (https://doi.org/10.1186/gb-2012-13-8-r67). This is utilized in order to provide better resultion of the actual cross-linking site which allows the determination of the exact nucleotide. In 2020 iCLIP2 was developed by Buchbender et al. (https://doi.org/10.1016/j.ymeth.2019.10.003) which provided several improvements:

- Bead based cDNA size selection instead of gel based 
    - Quicker, easier and cheaper
    
- Second adapter is ligated to 3' end directly instead of by circularization and relinearisation 
    - Quicker and reduces possibility of sample loss

- Longer barcode and UMI sequences. 6-nt barcodes and 9-nt UMIs instead of 4-nt barcodes and 5-nt UMIs 
    - higher barcode specificity
    - Lower chance of duplicate UMIs

- First PCR before size selection which is then followed by second PCR instead of single PCR after size selection 
    - cDNAs lost in isze selection are already pre-amplified and are thus present in the final library

.. figure:: images/iCLIP2-overview.png
    :width: 600
    :alt: Overview of the iCLIP2 method. Taken from: https://doi.org/10.1016/j.ymeth.2019.10.003
    
    Overview of the iCLIP2 method. Source: https://doi.org/10.1016/j.ymeth.2019.10.003

About barcodes and UMIs
-----------------------
Barcodes and UMIs (also called random barcodes) are nucleotide sequences that can be added to every read. Barcodes are experiment or sample specific and are used to differentiate reads from different experiments after they were multiplexed for sequencing. A certain nucleotide sequence is assigned to every experiment which will then be present at a distinct position within every read from that experiment. By :ref:`providing these nucleotide information <barcodes>` tools can assign all reads to their appropriate experiment.  
UMIs are nucleotide sequences consisting of random nucleotides. They are used in order to differentiate reads that were actually present multiple times and PCR duplictates. Since they are random, the possibility of 2 reads that are not PCR duplicates is extremely low given that a UMI of 9-nt provides 262144 different possible sequences. When 2 reads allign to the exact same position and also possess the same UMI they are considered PCR duplicates and therefore all but one are discarded.
Example of barcode sequences:

.. raw:: html

    <ul>
        <li><span style="color: red"> Random barcode (UMI)</span></li>
        <li><span style="color: green"> Experimental barcode</span></li>
        <li><span style="color: black"> Actual read</span></li>
    </ul>

Barcode example for iCLIP:

.. raw:: html

    <div style="font-family: monospace; background-color: #f0f0f0; padding: 10px;">
        <span style="color: red;">NNN</span>
        <span style="color: green;">XXXX</span>
        <span style="color: red;">NN</span>
        <span style="color: black;">AGTCGCTCGCTCGATAGCTAGCTAGCTAGCT</span>
    </div>
    <br>


Barcode example for iCLIP2:

.. raw:: html

    <div style="font-family: monospace; background-color: #f0f0f0; padding: 10px;">
        <span style="color: red;">NNNNN</span>
        <span style="color: green;">XXXXXX</span>
        <span style="color: red;">NNNN</span>
        <span style="color: black;">AGTCGCTCGCTCGATAGCTAGCTAGCTAGCT</span>
    </div>
    <br>

Information about how to use PARANOiD with different barcodes can be found :ref:`here <barcode-pattern>`.

About other CLIP methods
------------------------

eCLIP
^^^^^
eCLIP (enhanced CLIP) is an enhanced version of iCLIP that was published in `2016 by Van Nostrad et al. <https://doi.org/10.1038/nmeth.3810>`_. It offers several improvements in the library preparation such as performing adapter ligation in 2 distinct steps instead of cirularising and relinearising RNAs which was also improved in the iCLIP2 protocol. eCLIP offers single nucleotide resultion of cross-linking events while decreading the required PCR amplification, decreasing the amount of discarded reads during PCR deduplication. As the standard eCLIP protocol results in paired-end reads with read 1 containing the experimental barcode and read 2 the UMI together with the cross-link site information, resulting reads cannot be processed with PARANOiD. Here we propose `CLIP-explorer <https://doi.org/10.1093/gigascience/giaa108>`_ or `racoon-clip <https://doi.org/10.1093/bioadv/vbae084>`_ which are both able to handle reads generated by iCLIP as well as eCLIP experiments.

PAR-CLIP
^^^^^^^^

About iCLIP pipelines
---------------------

Besides PARANOiD several other iCLIP pipelines have been developed and published each having different strengths and weaknesses. Here we want to give a short overview of the most relevant ones and show the gap that PARANOiD fills out.

.. _CLIP-explorer:

CLIP-Explorer
^^^^^^^^^^^^^
`CLIP-explorer <https://doi.org/10.1093/gigascience/giaa108>`_ is a `Galaxy pipeline <https://clipseq.usegalaxy.eu/>`_ for iCLIP and eCLIP data which offers an easy usability even to users without any informatics background due to its implemenatation in Galaxy. It offers the implementation of several different peak callers, each implemented in an own workflow and a motif detection.  
Due to it's implementation in GALAXY datasets needs to be uploaded which can make it unsuited for potentially sensistive data. Furthermore, it's either necessary to create an account and login to run the workflow or host in on an own GALAXY server. The main workflow of CLIP-explorer is restricted to very few reference genomes with hg38 being the main focus, whereas PARANOiD offers the analysis with all reference genomes provided by users. While CLIP-explorer offers very detailed processing steps, the minimum requirements to perform appear higher than necessary without providing direct information for each required file. For example, input reads are required to be paired-end and an annotation file is required to run any of the available workflows. Lastly, demultiplexing of reads is provided as a separated workflow, making multiple execution steps necessary. 

.. _FAST-iCLIP:

FAST-iCLIP
^^^^^^^^^^
`FAST iCLIP <https://github.com/ChangLab/FAST-iCLIP>`_ is an iCLIP pipeline that is restricted to the analysis of human (GRCh38) and mouse (GRCm38). It relies on outdated dependencies, such as Python2, which are no longer maintained and may expose the system to potential security risks. Therefore, we would advise the usage of other pipelines in order to avoid the execution of deprecated software thus and the security risks.

