.. _section-analyses:

Included Analyses
=================

Short overview of all analyses implemented in PARANOiD

.. _basic-analysis:

Basic analysis
--------------

PARANOiD's basic analysis includes preprocessing of :ref:`FASTQ files <read-file>`, demultiplexing, aligning reads to a reference and calculating the cross-linking position based on the alignment. These positions are then translated into :ref:`WIG, BIGWIG and BEDGRAPH files <output-cross-link-sites-raw>` and given as output to the user. Additionally, a distribution of the :ref:`peak height <output-peak-height-distribution>` and the :ref:`strandness <output-strand-distribution>` are given as output together with a :ref:`statistics overview <output-statistics>` of important processes. Lastly, an :ref:`XML file <output-igv-session>` is generated that can be imported into `IGV <https://software.broadinstitute.org/software/igv/>` to automatically visualize results generated by PARANOiD.
The preprocessing involves adapter removal, quality filtering, splitting reads according to their experimental barcode, and removing the whole barcode. The alignment can be done via :ref:`two different alignment tools (Bowtie2 or STAR) <domain>` and is followed by a deduplication step in which PCR duplicates are removed. Finally, alignments are filtered via the :ref:`MAPQ score <mapq>` and cross-linking positions are calculated for each alignment. :ref:`two WIG, BIGWIG, and BEDGRAPH files <output-cross-link-sites-raw>` are generated for each sample - one for forward and one for reverse alignments. When analyzing peaks you should consider that iCLIP tends to have a bias towards cross-linking at Uridines (`Sugimoto et al. 2012 <10.1186/gb-2012-13-8-r67>`).

.. parsed-literal::
    Associated parameters (preprocessing)
    :ref:`-\-barcode_pattern <barcode-pattern>`           States composition of random and experimental barcodes
    :ref:`-\-barcode_mismatches <barcode-mismatches>`        Number of allowed mismatches within the experimental barcode to still align it to its sample
    :ref:`-\-omit_demultiplexing <parameters-omit-demultiplexing>`  Omits demultiplexing step
    :ref:`-\-min_length <min-read-length>`                Minimum length of reads required to be retained after adapter removal
    :ref:`-\-min_qual <min-base-qual>`                  Minimum quality a base needs to be retained
    :ref:`-\-min_percent_qual_filter <percent-qual-filter>`  Percentage of bases that must exceed the quality threshold to retain the read 

    Associated parameters (alignment & cross link site determination)
    :ref:`-\-domain <domain>`                   Specifies whether Bowtie2 or STAR is used as the aligner
    :ref:`-\-mapq <mapq>`                     Minimum MAPQ score required to retain alignments
    :ref:`-\-max_alignments <max-alignments>`          Maximum number of alignments reported by the mapping tool
    :ref:`-\-report_all_alignments <report-all-alignments>`    Reports all possible alignments (may be filtered out later)

.. _merge-replicates-analysis:

Merge replicates
----------------

Merges several replicates into a single representative dataset which can be used for publications, posters or presentations. 
This dataset shows the mean hit count at each position. Additionally, a correlation analysis is performed to evaluate the sample similarity and therefore provide a rationale for this analysis. Correlation is performed on raw cross-link sites
(or on significant ones in case :ref:`peak calling <peak-calling>` is used) via the Pearson correlation.
This option is deactivated by default.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-merge_replicates <merge-replicates>`            Merges replicates based on the name specified in the :ref:`barcode file <barcodes>`
    :ref:`-\-correlation_analysis <correlation-analysis>`       Performs a correlation analysis on merged replicates

.. _RNA-subtype-analysis:

RNA subtypes
------------

Analysis to determine whether the protein of interest tends to bind to specific RNA subtypes or regions. As this is determined 
via the :ref:`annotation file <annotation>`, only subtypes included in it can be determined (shown in column 3). 
To see which RNA subtypes are included in the annotation file, a :ref:`script <determine-feature-types>` was added. 
When choosing RNA subtypes one has to be careful not to use subtypes that are hierarchically higher or lower than each other as 
they will at least partially cover the same reference regions, which makes hits in these regions ambiguous. 
The `SO terms <https://github.com/The-Sequence-Ontology/SO-Ontologies/blob/master/Ontology_Files/subsets/SOFA.obo>`_ can 
be used to get an overview of the official hierarchical structures of annotation files. 
This analysis can only be performed when an :ref:`annotation file <annotation>` is provided.

.. parsed-literal::
    Associated parameters
    :ref:`-\-run_rna_subtype <rna-subtypes>`          Performs RNA subtype analysis
    :ref:`-\-gene_id <gene-id>`               Specifies the gene ID tag used in the :ref:`annotation file <annotation>`
    :ref:`-\-color_barplot <color-barplot>`        Specifies colors of the bars in the barplot generated by this analysisanalysis
    :ref:`-\-rna_subtypes <rna-subtypes>`          Specifies RNA subtypes/regions used for this analysis
    :ref:`-\-report_not_assigned <parameters-report-not-assigned>`  Reports not assigned peaks among RNA subtypes
    :ref:`-\-annotation_extension <parameters-annotation-extension>`    Extension of annotation file. Accepts GFF (for GFF3) or GTF
    :ref:`-\-show_ambiguous <parameters-show-ambiguous>`        Shows ambiguous peaks in normal subtype distribution

.. _transcript-analysis:

Transcript analysis
-------------------

Analysis to show if specific RNAs are more prone to interact with the protein of interest. 
If choosing this analysis, a file containing all RNAs of interest should be used as the input reference instead of the genome. 
All RNAs of interest (or artificial RNAs present in the sample) can be combined into a single FASTA file. If the general 
transcriptome of an organism is to be examined, they are often available alongside the genome and annotation of the organism. 
If not, a FASTA file containing the transcripts can be generated as follows (requires the genome and an annotation file):

'''
gffread -w output_transcripts.fa -g input_reference_genome.fa input_annotation.gff3
'''

.. parsed-literal::
    Associated parameters
    :ref:`-\-map_to_transcripts <map-to-transcripts>`             Activates transcript analysis
    :ref:`-\-number_top_transcripts <number-top-transcripts>`         Number of transcripts with most hits per sample that are offered as output

.. _peak-calling:

Peak calling
------------

Results obtained from analyzed iCLIP experiments typically contain a fair amount of background noise (signal not caused by
the actual protein-RNA interaction). This can be due to the reverse transcription not terminating when encountering an
amino acid or by a covalent binding of the protein of interest with an RNA just because they were in close proximity. Peak calling aims to filter out this background noise and thus reduce the amount of false-positive signal. 
PARANOiD employs `PureCLIP <https://github.com/skrakau/PureCLIP>`_ for its peak calling process. PureCLIP uses a hidden Markov model
to divide the reference into four different states based on the peak distribution (0-based). Additionally, identified peaks in close proximity 
can be merged into binding regions. Please note that in order to run PureCLIP all non ACGTN nucleotide letters need to be replaced with Ns.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-omit_peak_calling <omit-peak-calling>`                           Omits peak calling analysis
    :ref:`-\-peak_calling_for_high_coverage <peak-calling-for-high-coverage>`              Adds parameters to PureCLIP which can allow its successful execution for high coverage samples
    :ref:`-\-peak_calling_regions <peak-calling-regions>`                        Allows merging of several cross-link sites in close proximity to a cross link region
    :ref:`-\-peak_calling_regions_width <peak-calling-region-width>`                  Sets the distance within which cross-link sites in close proximity are allowed to be merged

.. _motif-detection:

Motif detection
---------------

Protein binding sites are often determined by protein-specific RNA motifs. These motifs are typically found at or in close proximity to cross-linking sites. To identify these motifs, motif detection was implemented. When omitting peak calling (:ref:`-\-omit_peak_calling <omit-peak-calling>`), background noise is filtered out by using only the top percentiles of cross-link peaks (by default only the top 10% are used: :ref:`-\-percentile 90 <percentile>`) in the same manner as in the :ref:`peak distance analysis <peak-distance-analysis>`. Sequences surrounding all peaks above the threshold are extracted and provided as output. All extracted sequences are then used for motif detection via `STREME <https://meme-suite.org/meme/doc/streme.html>`_, which returns several enriched motifs. When analyzing motifs you should consider that iCLIP tends to have a bias towards cross-linking at Uridines (`Sugimoto et al. 2012 <10.1186/gb-2012-13-8-r67>`) which may influence the resulting motifs. This can be bypassed with the option :ref:`-\-omit_cl_nucleotide <omit-cl-nucleotide>` which replaces the cross-linked nucleotide with an N.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-sequence_extraction <sequence-extraction>`          Performs sequence extraction and motif detection
    :ref:`-\-percentile <percentile>`                        Sets the threshold for peak values used in this analysis (in percentiles)
    :ref:`-\-seq_len <sequence-extraction-length>`                          Number of nucleotides extracted from each side of a cross-link site
    :ref:`-\-omit_cl_nucleotide <omit-cl-nucleotide>`                Omits the nucleotide at the cross link position
    :ref:`-\-omit_cl_width <omit-cl-width>`                     Omits the nucleotides surrounding the cross link position
    :ref:`-\-remove_overlaps <remove-overlaps>`                   Removes overlapping sequences 
    :ref:`-\-max_motif_num <max-number-of-motifs>`                    Specifies the maximum number of motifs to generate
    :ref:`-\-min_motif_width <min-motif-width>`                   Specifies the minimum width allowed for motifs
    :ref:`-\-max_motif_width <max-motif-width>`                   Specifies the maximum width allowed for motifs

.. _peak-distance-analysis:

Peak distance analysis
----------------------

Some proteins bind to long stretches of RNA instead of certain motif-dependent RNA subregions. This is, for example, the case with the Nucleocapsid (N) protein of several virus species which bind to a distinct number of nucleotides per N protein while packaging the viral RNA. The peak distance analysis was implemented to detect such periodical RNA-protein interactions by determining the occurences of distances between peaks. When omitting peak calling (:ref:`-\-omit_peak_calling <omit-peak-calling>`), background noise is filtered out by using only the top percentiles of peaks (by default only the top 10% are used: :ref:`-\-percentile 90 <percentile>`) in the same manner as in the :ref:`motif detection <motif-detection>`. Then, iterating over each peak, the distances to all other peaks which are within a specified distance (default 30 nt: :ref:`-\-distance 30 <distance>`) are measured, summarized and provided as output TSV file and barplot.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-peak_distance <peak-distance>`     Performs the peak distance analysis
    :ref:`-\-percentile <percentile>`             Sets the threshold for peak values used in this analysis (in percentiles)
    :ref:`-\-distance <distance>`             Maximum reported distance between peaks
