Included Analyses
=================

Short overview of all analyses implemented in PARANOiD

.. _basic-analysis:

Basic analysis
--------------

The basic analysis of PARANOiD includes the preprocessing of :ref:`FASTQ files <read-file>`, demultiplexing, aligning reads to a reference and calculating the cross-linking position based on the alignment. These positions are then translated into :ref:`WIG, BIGWIG and BEDGRAPH files <output-cross-link-sites-raw>` and given as output to the user. Additionally, a distribution of the :ref:`peak height <output-peak-height-distribution>` and the :ref:`stradnedness <output-strand-distribution>` are given as output together with a :ref:`statistics overview <output-statistics>` of important proceses. Lastly, an :ref:`XML file <output-igv-session>` is generated that can be importet by the `IGV <https://software.broadinstitute.org/software/igv/>` to automatically visualize results generated by PARANOiD.
The preprocessing involves adapter removal, quality filtering, splitting reads according to their experimental barcode and removing the whole barcode. The alignmentcan be done via :ref:`2 different alignment tools (Bowtie2 or STAR) <domain>` and is followed by a deduplication step in which PCR duplicates are removed. Finally, alignments are filtered via the :ref:`MAPQ score <mapq>` and cross-linking positions are calculated for each alignment. :ref:`2 WIG, BIGWIG and BEDGRAPH files <output-cross-link-sites-raw>` are generated for each sample - one for forward and one for reverse alignments. When analyzing peaks you should consider that iCLIP tends to have a bias towards cross-linking with Uridines (`Sugimoto et al. 2012 <10.1186/gb-2012-13-8-r67>`).

.. parsed-literal::
    Associated parameters (preprocessing)
    :ref:`-\-barcode_pattern <barcode-pattern>`           States composition of random and experimental barcodes
    :ref:`-\-barcode_mismatches <barcode-mismatches>`        Number of mismatches allowed mismatches within experimental barcode to still align it to it's sample
    :ref:`-\-min_length <min-read-length>`                Minimum length of reads necessary to retain after adapter removal
    :ref:`-\-min_qual <min-base-qual>`                  Minimum quality a base needs to retain
    :ref:`-\-min_percent_qual_filter <percent-qual-filter>`   Percentage of bases above the :ref:`quality threshold <min-base-qual>` necessary to retain the read 

    Associated parameters (alignment & cross link site determination)
    :ref:`-\-domain <domain>`                   States if Bowtie2 or STAR is being used as aligner
    :ref:`-\-mapq <mapq>`                     Minimum MAPQ score for alignments necessary to retain
    :ref:`-\-max_alignments <max-alignments>`           Maximum number of alignments provided by the mapping tool
    :ref:`-\-report_all_alignments <report-all-alignments>`    Reports all possible alignments (might be filtered out later on) 

.. _merge-replicates-analysis:

Merge replicates
----------------

Merges several replicates into a single representative version which can be used for publications, posters or presentations. 
This version shows the mean hit count for every position. Additionally, a correlation analysis is performed to give the user 
an evaluation of the sample similarity and therefore a rationale for this analysis. The correlation is performed on raw cross-link sites
(or on significant ones in case :ref:`peak calling <peak-calling>` is used) via the Pearson correlation.
Is deactivated by default.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-merge_replicates <merge-replicates>`           Merges replicates according to the name in the :ref:`barcode file <barcodes>`
    :ref:`-\-correlation_analysis <correlation-analysis>`       Does a correlation analysis for merged replicates

.. _RNA-subtype-analysis:

RNA subtypes
------------

Analysis to determine if the protein of interest is prone to bind to specific RNA subtypes or regions. As this is determined 
via the :ref:`annotation file <annotation>` only subtypes included there can be determined (shown in column 3). 
To see which RNA subtypes are included in the annotation file a :ref:`script <determine-feature-types>` was added. 
When choosing RNA subtypes one has to be careful not to use subtypes that are hierarchically higher or lower to each other as 
these will at least partially cover the same reference regions making hits in these regions ambiguous. 
The `SO ontologies <https://github.com/The-Sequence-Ontology/SO-Ontologies/blob/master/Ontology_Files/subsets/SOFA.obo>`_ can 
be used to get an overview of the official hierarchical structures of annotation files. 
Is activated when an :ref:`annotation file <annotation>` is provided.

.. parsed-literal::
    Associated parameters
    :ref:`-\-run_rna_subtype <rna-subtypes>`          Performs RNA subtype analysis
    :ref:`-\-gene_id <gene-id>`               Tag for the gene ID used within the :ref:`annotation file <annotation>`
    :ref:`-\-color_barplot <color-barplot>`         Color bars within the barplot generated by this analysis
    :ref:`-\-rna_subtypes <rna-subtypes>`          RNA subtypes/regions used for this analysis

.. _transcript-analysis:

Transcript analysis
-------------------

Analysis to show if specific RNAs are more prone to interact with the the protein of interest. 
If choosing this analysis a file containing all RNAs of interest should be used as input reference instead of the genome. 
Here all RNAs of interest (or artificial RNAs present in the sample) can be combined to a single fasta file. If the general 
transcriptome of an organism shall be examined, they can often be accessed next to the genome and annotation of the organism. 
If not a FASTA file containing the transcripts can be generated as follows (needs the genome and an annotation file):

'''
gffread -w output_transcripts.fa -g input_reference_genome.fa input_annotation.gff3
'''

.. parsed-literal::
    Associated parameters
    :ref:`-\-map_to_transcripts <map-to-transcripts>`             Activates transcript analysis
    :ref:`-\-number_top_transcripts <number-top-transcripts>`         Amount of transcripts with most hits per sample that are offered as output

.. _peak-calling:

Peak calling
------------

Results obtained from analyzed iCLIP experiments typically contain a fair amount of background noise (signal not caused by
the actual protein-RNA interaction). This can be due to the reverse transcription not terminating when encountering an
aminoacid or by a covalent binding of the protein of interest with an RNA just because their were in close proximity. Peak calling
is supposed to filter out this background noise and thus reduce the amount of false positive signal. 
PARANOiD employs `PureCLIP <https://github.com/skrakau/PureCLIP>`_ for its peak calling process. PureCLIP uses a hidden Markov model
to divide the reference into 4 different states based on the peak distribution (0-based). Additionally, identified peaks in close proximity 
can be merged into binding regions. Please note that in order to run PureCLIP all non ACGTN nucleotide letters need to be changed to Ns.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-omit_peak_calling <omit-peak-calling>`                           Omits peak calling analysis
    :ref:`-\-peak_calling_for_high_coverage <peak-calling-for-high-coverage>`              Adds parameters to PureCLIP which can allow it's succesful execution for high coverage samples
    :ref:`-\-peak_calling_regions <peak-calling-regions>`                        Allows merging several cross link sites in close proximity to a cross link region
    :ref:`-\-peak_calling_regions_width <peak-calling-region-width>`                  Sets the width until which cross link sites in close proximity are allowed to be merged

.. _motif-detection:

Motif detection
---------------

Protein binding sites are often determined by protein-specific RNA motifs. These motifs are typically found at or in close proximity to cross-linking sites. To identify these motifs the motif detection was implemented. When omitting peak calling (:ref:`-\-omit_peak_calling <omit-peak-calling>`) background noise is being filtered out by using only the top percentiles of peaks (by default only the top 10% are used: :ref:`-\-percentile 90 <percentile>`) in the same manner as in the :ref:`peak distance analysis <peak-distance-analysis>`. Sequences around all peaks above the threshold are extracted and provided as output. All extracted sequences are then used for motif detection via `streme <https://meme-suite.org/meme/doc/streme.html>`_, which offers several enriched sequences. When analyzing motifs you should consider that iCLIP tends to have a bias towards cross-linking with Uridines (`Sugimoto et al. 2012 <10.1186/gb-2012-13-8-r67>`) which could influence the resulting motifs. This can be bypassed with the option :ref:`-\-omit_cl_nucleotide <omit-cl-nucleotide>` which exchanges the cross-linked nucleotide with an N.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-sequence_extraction <sequence-extraction>`          Omits the sequence extraction and motif detection
    :ref:`-\-percentile <percentile>`                        Sets threshold for peak values used for this analysis using percentiles 
    :ref:`-\-seq_len <sequence-extraction-length>`                           Nucleotides extracted from each side of a cross link site
    :ref:`-\-omit_cl_nucleotide <omit-cl-nucleotide>`                Omits the nucleotide at the cross link position
    :ref:`-\-omit_cl_width <omit-cl-width>`                     Omits the nucleotides surrounding the cross link position
    :ref:`-\-remove_overlaps <remove-overlaps>`                   Removes overlapping sequences 
    :ref:`-\-max_motif_num <max-number-of-motifs>`                     Maximum number of motifs generated 
    :ref:`-\-min_motif_width <min-motif-width>`                   Minimum width allowed for motifs
    :ref:`-\-max_motif_width <max-motif-width>`                   Maximum width allowed for motifs 

.. _peak-distance-analysis:

Peak distance analysis
----------------------

Some proteins bind to long stretches of RNA instead of certain motif-dependent RNA subregions. This is, for example, the case with the Nucleocapsid (N) protein of several virus species which bind to a distinct number of nucleotides per N protein while packaging the viral RNA. The peak distance analysis was implemented to detect such periodical RNA-protein interactions by determining the occurences of distances between peaks. When omitting peak calling (:ref:`-\-omit_peak_calling <omit-peak-calling>`), background noise is being filtered out by using only the top percentiles of peaks (by default only the top 10% are used: :ref:`-\-percentile 90 <percentile>`) in the same manner as in the :ref:`motif detection <motif-detection>`. Then, going through every peak, the distances to all other peaks which are within a certain distance (by default 30 nt: :ref:`-\-distance 30 <distance>`) are measured, summarized and provided as a TSV file and visualized as a plot.

.. parsed-literal::
    Associated parameters:
    :ref:`-\-peak_distance <peak-distance>`     Omits the peak distance analysis
    :ref:`-\-percentile <percentile>`             Sets threshold for peak values used for this analysis using percentiles 