.. _section-parameters:

PARANOiD Parameters
===================

Explanation of all PARANOiD parameters

.. _parameters-read-file:

-\-reads
--------

| Essential parameter!
| Specifies the  :ref:`file <read-file>` containing reads obtained from iCLIP experiments.
| Expects a ``FASTQ`` file.

**Usage:**

.. code-block:: shell

    --reads /path/to/input-file.fastq


.. _parameters-barcodes:

-\-barcodes
-----------

| Essential parameter!
| Specifies the  :ref:`file <barcodes>` containing barcode sequences and experiment names. Required to split reads and assign them to their corresponding experiment.
| Expects a ``TSV`` file.

**Usage:**

.. code-block:: shell

    --barcodes /path/to/barcodes.tsv


.. _parameters-reference:

-\-reference
------------

| Essential parameter!
|  Specifies the :ref:`reference genome <reference>` used to align reads and determine the location of cross-link sites.
| Expects a ``FASTA`` file.

**Usage:**

.. code-block:: shell

    --reference /path/to/reference.fasta


.. _parameters-annotation:

-\-annotation
-------------

| States :ref:`annotation file <annotation>` used for the :ref:`RNA subtype analysis <RNA-subtype-analysis>`.
| Expects a ``GFF`` or ``GTF`` file.

**Usage:**

.. code-block:: shell

    --annotation /path/to/annotation.gff


.. _merge-replicates:

-\-merge_replicates
-------------------

Merges replicates into a single representative form. In order to do so experiment names must follow a specific naming convention which is further explained in the  :ref:`barcodes section <barcodes>`.


``Default: false``

**Usage:**

.. code-block:: shell

    --merge_replicates


.. _correlation-analysis:

-\-correlation_analysis
-----------------------

Only applies when :ref:`replicate merging <merge-replicates>` is chosen.
Performs a correlation analysis of replicates to show their similarity (and thus if they should be merged at all).
May cause excessive memory usage for large reference genomes 

``Default: false``

**Usage:**  

.. code-block:: shell

    --correlation_analysis


.. _barcode-pattern:

-\-barcode_pattern
------------------

Adjusts the barcode pattern to different protocols. Default protocol is `iCLIP2 <https://doi.org/10.1016/j.ymeth.2019.10.003>`_.
``N`` s represent the random barcode and ``X`` s the experimental barcode

Usage (default):

.. code-block:: shell

    --barcode_pattern NNNNNXXXXXXNNNN

Example for iCLIP1

.. code-block:: shell

    --barcode_pattern NNNXXXXNN


.. _domain:

-\-domain
---------

Choose between bowtie2 and STAR to be used to align reads to the reference sequence. Bowtie2 should be used for prokarytic organisms or transcript sequences while STAR should be used for eukaryotic organisms (or rather all splicing capable organisms) as STAR is splicing aware. If using STAR for splicing capable organisms it is highly recommended to provide an :ref:`annotation file <annotation>` file besides the reference.

| Options:
| pro -> Bowtie2 (default)
| eu  -> STAR

Usage (default):

.. code-block:: shell

    --domain pro 

.. _max-alignments:

-\-max_alignments
-----------------

Maximum number of alignments the mapping tool provides per read. It is not guaranteed that this many alignments are found per read.
If you want to find as many alignments as possible please use the parameter :ref:`--report_all_alignments <report-all-alignments>`

Usage (default):

.. code-block:: shell

    --max_alignments 1 

.. _report-all-alignments:

-\-report_all_alignments
------------------------

If used the mapping tools will report all alignments rather than a few. Overwrites the option :ref:`--max_alignments <max-alignments>`

**Usage:**

.. code-block:: shell

    --report_all_alignments

.. _output-dir:

-\-output
---------

Specifies the output directory generated by PARANOiD.

Usage (default):

.. code-block:: shell

    --output ./output


.. _min-read-length:

-\-min_length
-------------

Specifies the minimum length a read must have after adapter removal to be retained. Reads that become shorter during adapter removal will be filtered out.

Usage (default):

.. code-block:: shell

    --min_length 30


.. _min-base-qual:

-\-min_qual
-----------

Specifies the minimum base quality. All bases below that quality are cut off. 
The quality score (also known as Phred quality score) describes the certainty of correctness of the base and is typically calculated as follows with e being the error probability: :math:`Q-Score = -10log_\text{10}(e)`

+---------------------+-------------------+-----------------+
| Phred Quality score | Error probability | Accuracy        |
+=====================+===================+=================+
| 10                  | 10%               | 90%             |
+---------------------+-------------------+-----------------+
| 20                  | 1%                | 99%             |
+---------------------+-------------------+-----------------+
| 30                  | 0.1%              | 99.9%           |
+---------------------+-------------------+-----------------+
| 40                  | 0.01%             | 99.99%          |
+---------------------+-------------------+-----------------+


Usage (default):

.. code-block:: shell

    --min_qual 20


.. _percent-qual-filter:

-\-min_percent_qual_filter
--------------------------

Percentage of nucleotides that need to have a quality score above the chosen :ref:`minimum base quality <min-base-qual>`.
Reads with less nucleotides above the desired quality will be removed.

Usage (default):

.. code-block:: shell

    --min_percent_qual_filter 90


.. _barcode-mismatches:

-\-barcode_mismatches
---------------------

Number of mismatches allowed within the experimental barcode to still assign a read to an experiment.
Typically, experimental barcodes should be designed with a v of at least 3 to each other in order to allow one mismatch. 

Usage (default):

.. code-block:: shell

    --barcode_mismatches 1


.. _mapq:

-\-mapq
-------

Minimum alignment quality (mapq score) an alignment needs to retain. The interpretation of score values depends on the aligner specified via :ref:`--domain <domain>`.
All alignments with a mapq score below will be removed after the alignment step. 
Please note that these are just a short overview of the meaning of MAPQ scores and that they can be more complex than shown here when going into details.
the MAPQ score can be found in alignment files (SAM/BAM/CRAM) in column 5.

Usage (default):

.. code-block:: shell

    --mapq 2


Score meanings for Bowtie2 (--domain pro)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Apart from the description in the table a higher MAPQ score means less allowed mismatches (with difference of the base quality a mismatched nucleotide has)

+---------------------+--------------------------------------------------------------------------------------------------------------+
| MAPQ score          | Description                                                                                                  |
+=====================+==============================================================================================================+
| 0                   | All mappable reads                                                                                           |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 1                   | Multimapped reads that have the same alignment quality at different positions                                |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 2-39                | Multimapped reads that have one specific alignment with a better score than the other potential positions    |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 40                  | Reads mappable to only one position                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 42                  | Reads mappable to only one position with an almost perfect alignment (best possible MAPQ score in Bowtie2). |
+---------------------+--------------------------------------------------------------------------------------------------------------+

More information can be found `here <http://biofinysics.blogspot.com/2014/05/how-does-bowtie2-assign-mapq-scores.html>`_

Score meanings for STAR (--domain eu)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

+---------------------+--------------------------------------------------------------------------------------------------------------+
| MAPQ score          | Description                                                                                                  |
+=====================+==============================================================================================================+
| 0                   | Maps to 10 or more positions                                                                                 |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 1                   | Maps to 4-9 positions                                                                                        |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 2                   | Maps to 3 positions                                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 3                   | Maps to 2 positions                                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 255                 | Reads mappable to only one position. Best MAPQ score in STAR alignments.                                     |
+---------------------+--------------------------------------------------------------------------------------------------------------+

| The mapping quality MAPQ (column 5) is 255 for uniquely mapping reads, and  :math:`MAPQ score = int(-10log_\text{10}(1-1/[\text{number of positions the read maps to}]))` for multi-mapping reads. This scheme is the same as the one used by TopHat [...]
| Source: `Bowtie2 manual <https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf>`_

.. _map-to-transcripts:

-\-map_to_transcripts
---------------------

Use this option when transcript sequences are provided instead of a reference genome. Returns the transcripts with most hits from each sample. 
More information can be found :ref:`here <transcript-analysis>`

``Default: false``

**Usage:**

.. code-block:: shell

    --map_to_transcripts


.. _number-top-transcripts:

-\-number_top_transcripts
-------------------------

Specifies how many top-hit transcripts to retain per sample that are selected if parameter :ref:`--map_to_transcripts <map-to-transcripts>` was used.
Since selection is done per sample, the total number of reported transcripts may exceed this value.

Usage (default):

.. code-block:: shell

    --number_top_transcripts 10


.. _omit-peak-calling:

-\-omit_peak_calling
--------------------

If specified :ref:`peak calling <peak-calling>` will not be performed.
By default, peak calling is performed

**Usage:**

.. code-block:: shell

    --omit_peak_calling


.. _peak-calling-for-high-coverage:

-\-peak_calling_for_high_coverage
---------------------------------

Only has an effect if :ref:`peak calling <peak-calling>` is performed. 
Proteins covering the whole reference genome can cause problems for PureCLIP causing it to throw an error. 
Based on our experience, the parameters added by this option can improve PureCLIP with performing its analysis.
Adds the following parameters to the PureCLIP command: ``-mtc 5000 -mtc2 5000 -ld``

**Usage:**

.. code-block:: shell
    
    --peak_calling_for_high_coverage


.. _peak-calling-regions:

-\-peak_calling_regions
-----------------------

Takes effect only if :ref:`peak calling <peak-calling>` is enabled.
If specified, PureCLIP returns peak regions instead of individual peak sites.

**Usage:**

.. code-block:: shell

    --peak_calling_regions


.. _peak-calling-region-width:

-\-peak_calling_regions_width
-----------------------------

Takes effect only if :ref:`peak calling regions <peak-calling-regions>` is enabled.
Specifies the width of peak regions reported by PureCLIP.

Usage (default):

.. code-block:: shell

    --peak_calling_regions_width 8


.. _gene-id:

-\-gene_id
----------

| Only has an effect if an :ref:`annotation file <annotation>` is provided and thus the :ref:`RNA subtype analysis <RNA-subtype-analysis>` performed.
|Name of the tag used to identify gene IDs. Is found in the last column of annotation files, typically as the first tag-value pair.
| This column typically looks like the following: 
| ``ID=gene-LOC101842720;Dbxref=GeneID:101842720;Name=LOC101842720;gbkey=Gene;gene=LOC101842720;gene_biotype=pseudogene;pseudo=true``
|

In this case, the required tag is ``ID``.

Usage (default):

.. code-block:: shell

    --gene_id ID


.. _color-barplot:

-\-color_barplot
----------------

Specifies the color of bar plots generated by PARANOiD.
Applies to graphs generated in the following analyses: :ref:`peak height distribution <output-peak-height-distribution>`, :ref:`RNA subtype analysis <RNA-subtype-analysis>` and the :ref:`experimental barcode distribution <output-statistics>`.
Color is specified using a hexadecimal color code. If unsure which code corresponds to which color, websites like this `Example <https://www.color-hex.com/>` can help.

Usage (default):

.. code-block:: shell

    --color_barplot #69b3a2

.. _run-rna-subtypes:

-\-run_rna_subtype
-------------------

Enables the :ref:`RNA subtype analysis <RNA-subtype-analysis>`

**Usage:**

.. code-block:: shell

    --run_rna_subtype


.. _rna-subtypes:

-\-rna_subtypes
---------------

Takes effect only if an :ref:`annotation file <annotation>` is provided and :ref:`-\-run_rna_subtype <run-rna-subtypes>` is  enabled, triggering :ref:`RNA subtype analysis <RNA-subtype-analysis>`. 
Specifies which RNA subtypes (or regions) to include in the :ref:`RNA subtype analysis <RNA-subtype-analysis>`. 
Subtypes must be comma-separated and must appear in the feature type column (3rd column) of the :ref:`annotation file <annotation>`. If these conditions are not met, the analysis may fail or is performed incorrectly. 
If not sure which RNA subtypes are included within your annotation file you can use the script :ref:`featuretypes-from-gtfgff.awk <determine-feature-types>`. 
"Avoid selecting subtypes or regions that are hierarchically related, as they may overlap and cause peaks to appear as **ambiguous**. Information about the hierarchical structure of RNA subtypes/regions can be obtained `here <https://github.com/The-Sequence-Ontology/SO-Ontologies/blob/master/Ontology_Files/subsets/SOFA.obo/>`_.

Usage (default):

.. code-block:: shell

    --rna_subtypes 3_prime_UTR,transcript,5_prime_UTR


.. _peak-distance:

-\-peak_distance
---------------------

Enables the :ref:`peak distance analysis <peak-distance-analysis>` step.

**Usage:**

.. code-block:: shell

    --peak_distance


.. _distance:

-\-distance
-----------------------------------

Maximum allowed distance between peaks for the :ref:`peak distance analysis <peak-distance-analysis>`. 

Usage (default):

.. code-block:: shell

    --distance 30


.. _percentile:

-\-percentile
-------------

Peak percentiles for :ref:`peak distance analysis <peak-distance-analysis>` and :ref:`sequence extraction/motif analysis <motif-detection>`. Only peaks with values above this threshold are considered; all others are treated as background noise and ignored.
For example, a percentile value of 90 includes only the top 10% of peaks.
Only applies when :ref:`peak calling is omitted <omit-peak-calling>`.

Usage (default):

.. code-block:: shell

    --percentile 90


.. _sequence-extraction:

-\-sequence_extraction
---------------------------

Omits the :ref:`motif detection <motif-detection>` step.

**Usage:**

.. code-block:: shell

    --sequence_extraction


.. _sequence-extraction-length:

-\-seq_len
----------

Only applies if :ref:`motif detection <motif-detection>` is enabled.
Length in nucleotides to each side of a peak that is extracted from the :ref:`reference <reference>`.  
A value of 20 will lead to sequences of 41 nucleotides being extracted. (i.e. 20 nt upstream + 1 cross-link nucleotide + 20 nt downstream)

Usage (default):

.. code-block:: shell

    --seq_len 20


.. _omit-cl-nucleotide:

-\-omit_cl_nucleotide
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
The nucleotide at the cross-link site will be replaced with an **N** during sequence extraction.
This can improve motif detection, as iCLIP protocols often show a uridine (U) bias at cross-link sites.

**Usage:**

.. code-block:: shell
    
    --omit_cl_nucleotide


.. _omit-cl-width:

-\-omit_cl_width
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed and the :ref:`cl nucleotide is omitted <omit-cl-nucleotide>`.
Replaces nucleotides flanking the cross-link site with **N** to reduce artifacts from uridine-rich regions.
The value defines how many nucleotides upstream and downstream of the cross-link site are masked.

Usage (default):

.. code-block:: shell
    
    --omit_cl_width 0


.. _remove-overlaps:

-\-remove_overlaps
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed. 
Removes cross-link sites with lower peak values if their extracted sequences would overlap with those of neighboring sites.
This can be done to avoid doubled sequences during motif detection.

**Usage:**

.. code-block:: shell
    
    --remove_overlaps


.. _max-number-of-motifs:

-\-max_motif_num
----------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Maximum number of motifs reported by STREME.

Usage (default):

.. code-block:: shell
    
    --max_motif_num 50


.. _min-motif-width:

-\-min_motif_width
------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Minimum length of motifs reported by STREME.
Must be at least 3.

Usage (default):

.. code-block:: shell
    
    --min_motif_width 8


.. _max-motif-width:

-\-max_motif_width
------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Maximum length of motifs reported by STREME.
Must not exceed 30.

Usage (default):

.. code-block:: shell

    --max_motif_width 15
