.. _section-parameters:

PARANOiD Parameters
===================

Explanation of all PARANOiD parameters

.. _parameters-read-file:

-\-reads
--------

| Essential parameter!
| States :ref:`file <read-file>` containing reads obtained by iCLIP experiments.
| Expects a ``FASTQ`` file.

Usage:

.. code-block:: shell

    --reads /path/to/input-file.fastq


.. _parameters-barcodes:

-\-barcodes
-----------

| Essential parameter!
| States :ref:`file <barcodes>` containing barcode sequences and experiment names. Necessary to split reads and allocate them to their experiment.
| Expects a ``TSV`` file.

Usage:

.. code-block:: shell

    --barcodes /path/to/barcodes.tsv


.. _parameters-reference:

-\-reference
------------

| Essential parameter!
| States :ref:`reference genome <reference>` used to align reads to and thus to determine the location of cross-link sites.
| Expects a ``FASTA`` file.

Usage:

.. code-block:: shell

    --reference /path/to/reference.fasta


.. _parameters-annotation:

-\-annotation
-------------

| States :ref:`annotation file <annotation>` used for the :ref:`RNA subtype analysius <RNA-subtype-analysis>`.
| Expects a ``GFF`` or ``GTF`` file.

Usage:

.. code-block:: shell

    --annotation /path/to/annotation.gff


.. _merge-replicates:

-\-merge_replicates
-------------------

Merges replicates into a single representatiove form. In order to do so experiment names need to named in a particular manner which is further explained in the  :ref:`barcodes section <barcodes>`.


``Default: false``

Usage:

.. code-block:: shell

    --merge_replicates


.. _correlation-analysis:

-\-correlation_analysis
-----------------------

Only applies when :ref:`replicate merging <merge-replicates>` is chosen.
Does a correlation analysis of replicates to show their similarity (and thus if they should be merged at all).
Can cause problems with large reference genomes due to excessive RAM usage.

``Default: false``

Usage:  

.. code-block:: shell

    --correlation_analysis


.. _barcode-pattern:

-\-barcode_pattern
------------------

Adapt barcode patterns to different protocols. Default protocol is `iCLIP2 <https://doi.org/10.1016/j.ymeth.2019.10.003>`_.
``N`` s represent the random barcode and ``X`` s the experimental barcode

Usage (default):

.. code-block:: shell

    --barcode_pattern NNNNNXXXXXXNNNN

Example for iCLIP1

.. code-block:: shell

    --barcode_pattern NNNXXXXNN


.. _domain:

-\-domain
---------

Choose between bowtie2 and STAR to be used to align reads to the reference sequence. Bowtie2 should be used for prokarytic organisms or transcript sequences while STAR should be used for eukaryotic organisms (or rather all splicing capable organisms) as STAR is splicing aware. If using STAR for splicing capable organisms it is highly recommended to provide an :ref:`annotation file <annotation>` file besides the reference.

| Options:
| pro -> Bowtie2 (default)
| eu  -> STAR

Usage (default):

.. code-block:: shell

    --domain pro 

.. _max-alignments:

-\-max_alignments
-----------------

Maximum number of alignments the mapping tool provides per read. It is not guaranteed that this many alignments are found per read.
If you want to find as many alignments as possible please use the parameter :ref:`--report_all_alignments <report-all-alignments>`

Usage (default):

.. code-block:: shell

    --max_alignments 1 

.. _report-all-alignments:

-\-report_all_alignments
------------------------

If used the mapping tools will report all alignments rather than a few. Overwrites the option :ref:`--max_alignments <max-alignments>`

Usage:

.. code-block:: shell

    --report_all_alignments

.. _output-dir:

-\-output
---------

Specify directory to which output generated by PARANOiD will be written.

Usage (default):

.. code-block:: shell

    --output ./output


.. _min-read-length:

-\-min_length
-------------

Specify minimum length a read needs to have after adapter removal to persist. Reads that become shorter during adapter removal will be filtered out.

Usage (default):

.. code-block:: shell

    --min_length 30


.. _min-base-qual:

-\-min_qual
-----------

Minimum quality for bases. All bases below that quality are cut off. 
The quality score (also known as Phred quality score) describes the certainty of correctness of the base and is typically calculated as follows with e being the error probability: :math:`Q-Score = -10log_\text{10}(e)`

+---------------------+-------------------+-----------------+
| Phred Quality score | Error probability | Accuracy        |
+=====================+===================+=================+
| 10                  | 10%               | 90%             |
+---------------------+-------------------+-----------------+
| 20                  | 1%                | 99%             |
+---------------------+-------------------+-----------------+
| 30                  | 0.1%              | 99.9%           |
+---------------------+-------------------+-----------------+
| 40                  | 0.01%             | 99.99%          |
+---------------------+-------------------+-----------------+


Usage (default):

.. code-block:: shell

    --min_qual 20


.. _percent-qual-filter:

-\-min_percent_qual_filter
--------------------------

Percentage of nucleotides that need to have a quality score above the chosen :ref:`minimum base quality <min-base-qual>`.
Reads with less nucleotides above the desired quality will be removed.

Usage (default):

.. code-block:: shell

    --min_percent_qual_filter 90


.. _barcode-mismatches:

-\-barcode_mismatches
---------------------

Number of mismatches allowed within the experimental barcode to still assign a read to an experiment.
Typically, experimental barcodes should be designed with a v of at least 3 to each other in order to allow one mismatch. 

Usage (default):

.. code-block:: shell

    --barcode_mismatches 1


.. _mapq:

-\-mapq
-------

Minimum alignment quality (mapq score) an alignment needs to retain. The meaning of different scores is dependant on the aligner chosen via :ref:`--domain <domain>`.
All alignments with a mapq score below will be removed after the alignment step. 
Please note that these are just a short overview of the meaning of MAPQ scores and that they can be more complex than shown here when going into details.
the MAPQ score can be found in alignment files (SAM/BAM/CRAM) in column 5.

Usage (default):

.. code-block:: shell

    --mapq 2


Score meanings for Bowtie2 (--domain pro)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Apart from the description in the table a higher MAPQ score means less allowed mismatches (with difference of the base quality a mismatched nucleotide has)

+---------------------+--------------------------------------------------------------------------------------------------------------+
| MAPQ score          | Description                                                                                                  |
+=====================+==============================================================================================================+
| 0                   | All mappable reads                                                                                           |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 1                   | Multimapped reads that have the same alignment quality at different positions                                |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 2-39                | Mulitmapped reads that have one specific alignment with a better score than the other potential positions    |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 40                  | Reads mappable to only one position                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 42                  | Reads mappable to only one position with an almost perfect alignment. Best MAPQ score in Bowtie2 alignments  |
+---------------------+--------------------------------------------------------------------------------------------------------------+

More information can be found `here <http://biofinysics.blogspot.com/2014/05/how-does-bowtie2-assign-mapq-scores.html>`_

Score meanings for STAR (--domain eu)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

+---------------------+--------------------------------------------------------------------------------------------------------------+
| MAPQ score          | Description                                                                                                  |
+=====================+==============================================================================================================+
| 0                   | Maps to 10 or more positions                                                                                 |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 1                   | Maps to 4-9 positions                                                                                        |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 2                   | Maps to 3 positions                                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 3                   | Maps to 2 positions                                                                                          |
+---------------------+--------------------------------------------------------------------------------------------------------------+
| 255                 | Reads mappable to only one position. Best MAPQ score in STAR alignments.                                     |
+---------------------+--------------------------------------------------------------------------------------------------------------+

| The mapping quality MAPQ (column 5) is 255 for uniquely mapping reads, and  :math:`MAPQ score = int(-10log_\text{10}(1-1/[\text{number of positions the read maps to}]))` for multi-mapping reads. This scheme is same as the one used by TopHat [...]
| Source: `Bowtie2 manual <https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf>`_

.. _map-to-transcripts:

-\-map_to_transcripts
---------------------

Should be used when transcripts are given as reference instead of a reference genome. Returns the transcripts with most hits from each sample. 
More information can be found :ref:`here <transcript-analysis>`

``Default: false``

Usage:

.. code-block:: shell

    --map_to_transcripts


.. _number-top-transcripts:

-\-number_top_transcripts
-------------------------

The number of transcripts with most hits that are selected from each sample if parameter :ref:`--map_to_transcripts <map-to-transcripts>` was used.
As the amount is chosen from each sample the total number of transcripts can excede this number.

Usage (default):

.. code-block:: shell

    --number_top_transcripts 10


.. _omit-peak-calling:

-\-omit_peak_calling
--------------------

If specified :ref:`peak calling <peak-calling>` will not be performed.
Will be performed by default.

Usage:

.. code-block:: shell

    --omit_peak_calling


.. _peak-calling-for-high-coverage:

-\-peak_calling_for_high_coverage
---------------------------------

Only has an effect if :ref:`peak calling <peak-calling>` is performed. 
Proteins covering the whole reference genome can cause problems for PureCLIP causing it to throw an error. 
From our experience the parameters added by this argument can help PureCLIP with performing it's analysis.
Adds following arguments to the PureCLIP execution: ``-mtc 5000 -mtc2 5000 -ld``

Usage:

.. code-block:: shell
    
    --peak_calling_for_high_coverage


.. _peak-calling-regions:

-\-peak_calling_regions
-----------------------

Only has an effect if :ref:`peak calling <peak-calling>` is performed.
If specified peak regions instead of single peaks will be returned by PureCLIP.

Usage:

.. code-block:: shell

    --peak_calling_regions


.. _peak-calling-region-width:

-\-peak_calling_regions_width
-----------------------------

Only has an effect if :ref:`peak calling regions <peak-calling-regions>` are stated.
Changes the width of peak calling regions returned by PureCLIP.

Usage (default):

.. code-block:: shell

    --peak_calling_regions_width 8


.. _gene-id:

-\-gene_id
----------

| Only has an effect if an :ref:`annotation file <annotation>` is provided and thus the :ref:`RNA subtype analysis <RNA-subtype-analysis>` performed.
| Wording of the tag that describes the gene ID. Is found in the last column of annotation files, typically as the first tag-value pair.
| The column looks similar to this: 
| ``ID=gene-LOC101842720;Dbxref=GeneID:101842720;Name=LOC101842720;gbkey=Gene;gene=LOC101842720;gene_biotype=pseudogene;pseudo=true``
| In this case the tag necessary is `ID`.

Usage (default):

.. code-block:: shell

    --gene_id ID


.. _color-barplot:

-\-color_barplot
----------------

Color of barplots returned by PARANOiD.
Affects graphs generated by :ref:`peak height distribution <output-peak-height-distribution>`, :ref:`RNA subtype analysis <RNA-subtype-analysis>` and the :ref:`experimental barcode distribution <output-statistics>`.
Color is staded via a hexadecimal color code. If unsure which code translates to which color several websites can help to pick the correct one. `Example <https://www.color-hex.com/>`_

Usage (default):

.. code-block:: shell

    --color_barplot #69b3a2

.. _run-rna-subtypes:

-\-run_rna_subtype
-------------------

Performs the :ref:`RNA subtype analysis <RNA-subtype-analysis>`

Usage:

.. code-block:: shell

    --run_rna_subtype


.. _rna-subtypes:

-\-rna_subtypes
---------------

Only has an effect if an :ref:`annotation file <annotation>` is provided and :ref:`-\-run_rna_subtype <run-rna-subtypes>` is  chosen, performing :ref:`RNA subtype analysis <RNA-subtype-analysis>`. RNA subtypes/regions that shall be included in the :ref:`RNA subtype analysis <RNA-subtype-analysis>`. RNA subtypes need to be separated by a `,` and should appear in the :ref:`annotation file <annotation>` within the **feature type** column (3rd column). If both requirements are not met the analysis will either not be performed correctly or be aborted. If not sure which RNA subtypes are included within your annotation file you can use the script :ref:`featuretypes-from-gtfgff.awk <determine-feature-types>`. Additionally, users should beware not to choose subtypes/regions that are in a hierarchical relationship to each other as they can cover the same regions and thus make affected peaks appear as **ambiguous**. Inormation about the hierarchical structure of RNA subtypes/regions can be obtained `here <https://github.com/The-Sequence-Ontology/SO-Ontologies/blob/master/Ontology_Files/subsets/SOFA.obo/>`_.

Usage (default):

.. code-block:: shell

    --rna_subtypes 3_prime_UTR,transcript,5_prime_UTR


.. _omit-peak-distance:

-\-omit_peak_distance
---------------------

Omits the :ref:`peak distance analysis <peak-distance-analysis>`

Usage:

.. code-block:: shell

    --omit_peak_distance


.. _distance:

-\-distance
-----------------------------------

Max distance used for the :ref:`peak distance analysis <peak-distance-analysis>`. 

Usage (default):

.. code-block:: shell

    --distance 30


.. _percentile:

-\-percentile
-------------

Peak percentiles for :ref:`peak distance analysis <peak-distance-analysis>` and :ref:`sequence extraction/motif analysis <motif-detection>`. Only peaks with a value above this threshold are considered while all peaks below are omitted as background noise. 
A percentile of 90 means that only top 10% of peaks are used.  
Is only used when :ref:`peak calling is omitted <omit-peak-calling>`.

Usage (default):

.. code-block:: shell

    --percentile 90


.. _omit-sequence-extraction:

-\-omit_sequence_extraction
---------------------------

Omits the :ref:`motif detection <motif-detection>`

Usage:

.. code-block:: shell

    --omit_sequence_extraction


.. _sequence-extraction-length:

-\-seq_len
----------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Length in nucleotides to each side of a peak that is extracted from the :ref:`reference <reference>`.  
A value of 20 will lead to sequences of 41 nucleotides being extracted. (20nt upstream;cross-link nt;20nt downstream)

Usage (default):

.. code-block:: shell

    --seq_len 20


.. _omit-cl-nucleotide:

-\-omit_cl_nucleotide
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
The nucleotide directly at the cross-linking position will be substituted with an **N** when extracting sequences.  
Can improve the motif detection since iCLIP tends to have a bias towards **U** when cross-linking which can influence the motif search.

Usage:

.. code-block:: shell
    
    --omit_cl_nucleotide


.. _omit-cl-width:

-\-omit_cl_width
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed and the :ref:`cl nucleotide is omitted <omit-cl-nucleotide>`.
Omits nucleotides on both sides of the cross-linking position with an **N** to avoid potential uridine-polymers which can negatively influence the motif search. The number determines the amount of nucleotides on both sides that are to be replaced.

Usage (default):

.. code-block:: shell
    
    --omit_cl_width 0


.. _remove-overlaps:

-\-remove_overlaps
---------------------

Only applies when :ref:`motif detection <motif-detection>` is performed. 
Removes cross-link sites with lower peak values if their extracted sequence would overlap with the sequence from another cross-link site.
This can be done to avoid doubled sequences during motif detection.

Usage:

.. code-block:: shell
    
    --remove_overlaps


.. _max-number-of-motifs:

-\-max_motif_num
----------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Maximum number of motifs that is reported by streme.

Usage (default):

.. code-block:: shell
    
    --max_motif_num 50


.. _min-motif-width:

-\-min_motif_width
------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Minimum length of motifs reported by streme.
Cannot be lower than 3

Usage (default):

.. code-block:: shell
    
    --min_motif_width 8


.. _max-motif-width:

-\-max_motif_width
------------------

Only applies when :ref:`motif detection <motif-detection>` is performed.
Maximum length of motifs reported by streme.
Cannot be higher than 30

Usage (default):

.. code-block:: shell

    --max_motif_width 15
