<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>PARANOiD Inputs &mdash; PARANOiD  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PARANOiD Parameters" href="PARAMETERS.html" />
    <link rel="prev" title="Welcome to PARANOiD’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> PARANOiD
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Inputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#reads">Reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="#barcodes">Barcodes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotation">Annotation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="PARAMETERS.html">Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="ANALYSES.html">Analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="OUTPUT.html">Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PARANOiD</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">PARANOiD Inputs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/INPUT.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="paranoid-inputs">
<h1>PARANOiD Inputs<a class="headerlink" href="#paranoid-inputs" title="Permalink to this heading"></a></h1>
<p>Detailed description of all input files</p>
<div class="section" id="reads">
<span id="read-file"></span><h2>Reads<a class="headerlink" href="#reads" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">FASTQ</span></code> file containing all reads. Each read is represented by 4 lines:</p>
<ol class="arabic simple">
<li><p><strong>Sequence identifier</strong> and optional description. Starts with a <code class="docutils literal notranslate"><span class="pre">&#64;</span></code></p></li>
<li><p>Actual <strong>nucleotide sequence</strong> of the read</p></li>
<li><p><strong>Delimiter line</strong>. Starts with a <code class="docutils literal notranslate"><span class="pre">+</span></code></p></li>
<li><p><strong>Quality values</strong> of nucleotide sequence (line2). Must contain same number of symbols as line 2</p></li>
</ol>
<p>Example:</p>
<ol class="arabic simple">
<li><p>&#64;SEQ_ID</p></li>
<li><p>GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</p></li>
<li><p>+</p></li>
<li><p>!’’<em>((((***+))%%%++)(%%%%).1**</em>-+*’’))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</p></li>
</ol>
</div>
<div class="section" id="barcodes">
<span id="id1"></span><h2>Barcodes<a class="headerlink" href="#barcodes" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">TSV</span></code> file containing experiment names and the corresponding barcode sequences. Reads from the input <code class="docutils literal notranslate"><span class="pre">FASTQ</span></code> file are split according to the detected barcode sequence and assigned to the appropriate experiment. This results in one <code class="docutils literal notranslate"><span class="pre">FASTQ</span></code> file per experiment.</div>
<div class="line">When choosing the option to merge replicates <a class="reference internal" href="PARAMETERS.html#merge-replicates"><span class="std std-ref">[–merge_replicates]</span></a> the experiment names have to be chosen appropriately indicating which experiments belong together. In order to do that the appendix <code class="docutils literal notranslate"><span class="pre">_rep_&lt;number&gt;</span></code> has to be added to the experiment names, exchanging <code class="docutils literal notranslate"><span class="pre">&lt;number&gt;</span></code> with the replicate number.</div>
<div class="line">Barcode files consist of 2 columns separated by a single <code class="docutils literal notranslate"><span class="pre">tab</span></code>:</div>
</div>
<ol class="arabic simple">
<li><p>experiment name</p></li>
<li><p>barcode sequence present in reads</p></li>
</ol>
<div class="line-block">
<div class="line">Experiment names should only consist of Letters <code class="docutils literal notranslate"><span class="pre">{a-zA-Z}</span></code>, numbers <code class="docutils literal notranslate"><span class="pre">{1-9}</span></code> and underscores <code class="docutils literal notranslate"><span class="pre">_</span></code>. Any whitespaces (e.g. space, tab) will result in errors and thus the termination of the pipeline execution. The length of the barcode sequence is dependant on the protocol used an can be adapted via <a class="reference internal" href="PARAMETERS.html#barcode-pattern"><span class="std std-ref">[–barcode_pattern]</span></a>.</div>
</div>
<p>Example:</p>
<div class="line-block">
<div class="line">knockdown_N_rep_1    TGATAG</div>
<div class="line">knockdown_N_rep_2    AGTGGA</div>
<div class="line">knockdown_N_rep_3    GCTCGA</div>
<div class="line">mock_N_rep_1    TAAGTA</div>
<div class="line">mock_N_rep_2    GCAGTC</div>
<div class="line">mock_N_rep_3    CCTAGG</div>
</div>
</div>
<div class="section" id="reference">
<span id="id2"></span><h2>Reference<a class="headerlink" href="#reference" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">FASTA</span></code> file containing nucleotide data of interest. Is used to align reads to and thus find the location of cross-link sites. Can contain genomic or transcriptomic sequences of an organism or completely artificial sequences.
Every sequence consists of at least 2 lines:
1. Header
2-n. Nucleotide sequence
The header starts with a <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> and is followed by a description of the sequence
The sequence consists of nucleotides <code class="docutils literal notranslate"><span class="pre">{ACGTN}</span></code> and can span an arbitrary amount of lines</p>
<p>Example:</p>
<p>&gt;NW_024429180.1 Mesocricetus auratus isolate SY011 unplaced genomic scaffold
AACTCTGTTGtaaaaaggctttcccacattcattcCATTCATAAGGTTTCTGTACATTATGGATTCTTTCATGCCTTTTA
AGATGATTATGATATACATAGACTTTAACACCTCAAGAATAttcaggtttctctccagtatgacaATTTGGTCTAATTAT
AAAGAAGAATCAGATATTAAGGTTTTATCACTGTTTACACTCATGCTGTTCCCCTTCATTAAGGTTGGTTTGGATCTTTG
AATATACCTGGGTTCCTATAGTCTCCACCATCACATCTTTATGGAGATTCTTCTGGGAGGGATCCAGCAAATCCCACTCT
...</p>
</div>
<div class="section" id="annotation">
<span id="id3"></span><h2>Annotation<a class="headerlink" href="#annotation" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">GFF</span></code> or <code class="docutils literal notranslate"><span class="pre">GTF</span></code> file. Contains annotation information belonging to the reference used in the input. Describes features and their positions. PARANOiD does not rely on the annotation for it’s analysis, however it is highly recommended to provid it when working with splicing capable organisms (<a class="reference internal" href="PARAMETERS.html#domain"><span class="std std-ref">[–domain eu]</span></a>) as annotation files typically contain information about intron-exon structures which highly improve the mapping capability.
Furthermore, providing an annotation file enables the <a class="reference internal" href="ANALYSES.html#rna-subtype"><span class="std std-ref">RNA subtype analysis</span></a>.
Consists of several header lines followed by one line feature.
Header lines start with a <code class="docutils literal notranslate"><span class="pre">#</span></code> and contain general information about the annotation.</p>
<div class="line-block">
<div class="line">Feature lines consist of 9 columns which are separated by tabs:</div>
<div class="line">1. <strong>seqname</strong>: name of the chromosome or scaffold on which the feature is located</div>
<div class="line">2. <strong>source</strong>: name of the program that generated this feature or the source</div>
<div class="line">3. <strong>feature type</strong>: type of the current feature - e.g. exon - intron - CDS - mRNA - 3_prime_UTR - transcript - 5_prime_UTR</div>
<div class="line">4. <strong>start</strong>: Start position of the feature (1-based)</div>
<div class="line">5. <strong>end</strong>: End position of the feature (1-based)</div>
<div class="line">6. <strong>score</strong>: Float point value (can also simply be a <code class="docutils literal notranslate"><span class="pre">.</span></code>)</div>
<div class="line">7. <strong>strand</strong>: Stradn on which the feature is present. <code class="docutils literal notranslate"><span class="pre">+</span></code> for forward; <code class="docutils literal notranslate"><span class="pre">-</span></code> for reverse</div>
<div class="line">8. <strong>frame</strong>: Indicates which base of the feature is actually the first base of a codon. 0 -&gt; the first base of the feature is the first abse of a codon; 1 -&gt; the second base of the feature is the first base of a codon …. (can slo simply be a <code class="docutils literal notranslate"><span class="pre">.</span></code>)</div>
<div class="line">9. <strong>attributes</strong>: Semicolon separated list of tag-value pairs providing additional information</div>
</div>
<p>Example:</p>
<div class="line-block">
<div class="line">##gff-version 3</div>
<div class="line">#!gff-spec-version 1.21</div>
<div class="line">#!processor NCBI annotwriter</div>
<div class="line">#!genome-build BCM_Maur_2.0</div>
<div class="line">#!genome-build-accession NCBI_Assembly:GCF_017639785.1</div>
<div class="line">#!annotation-source NCBI Mesocricetus auratus Annotation Release 103</div>
<div class="line">##sequence-region NW_024429180.1 1 52462669</div>
<div class="line">##species <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=10036">https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=10036</a></div>
<div class="line">NW_024429180.1        RefSeq  region  1       52462669        .       +       .       ID=NW_024429180.1:1..52462669;Dbxref=taxon:10036;Name=Unknown;chromosome=Unknown;dev-stage=adult;gbkey=Src;genome=genomic;isolate=SY011;mol_type=genomic DNA;sex=female;tissue-type=liver</div>
<div class="line">NW_024429180.1        Gnomon  pseudogene      37366   38359   .       +       .       ID=gene-LOC101842720;Dbxref=GeneID:101842720;Name=LOC101842720;gbkey=Gene;gene=LOC101842720;gene_biotype=pseudogene;pseudo=true</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to PARANOiD’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="PARAMETERS.html" class="btn btn-neutral float-right" title="PARANOiD Parameters" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Patrick Barth.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>