<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>PARANOiD Outputs &mdash; PARANOiD  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Container usage" href="CONTAINER.html" />
    <link rel="prev" title="Included Analyses" href="ANALYSES.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> PARANOiD
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="INPUT.html">Inputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="BASICUSAGE.html">Basic Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="EXAMPLE.html">Example run</a></li>
<li class="toctree-l1"><a class="reference internal" href="PARAMETERS.html">Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="ANALYSES.html">Analyses</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#alignments">Alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cross-link-sites-peak-called">Cross link sites peak called</a></li>
<li class="toctree-l2"><a class="reference internal" href="#raw-cross-link-sites">Raw cross link sites</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#wig-wiggle">WIG (Wiggle)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bigwig">BIGWIG</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedgraph">BEDGRAPH</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization-with-igv">Visualization with IGV</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cross-link-sites-merged">Cross link sites merged</a></li>
<li class="toctree-l2"><a class="reference internal" href="#execution-metrics">Execution metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#igv-session">IGV-session</a></li>
<li class="toctree-l2"><a class="reference internal" href="#peak-height-distribution">Peak height distribution</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a></li>
<li class="toctree-l2"><a class="reference internal" href="#statistics">Statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#strand-distribution">Strand distribution</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CONTAINER.html">Container Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="CLUSTER.html">Cluster Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="ADDITIONAL_SCRIPTS.html">Additional scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="REQUIREMENTS.html">Requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PARANOiD</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">PARANOiD Outputs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/OUTPUT.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="paranoid-outputs">
<h1>PARANOiD Outputs<a class="headerlink" href="#paranoid-outputs" title="Permalink to this heading"></a></h1>
<p>Explanation of all outputs generated by PARANOiD</p>
<div class="section" id="alignments">
<span id="output-alignments"></span><h2>Alignments<a class="headerlink" href="#alignments" title="Permalink to this heading"></a></h2>
<p>Directory that contains deduplicated alignments in <code class="docutils literal notranslate"><span class="pre">BAM</span></code> format together with an index file in <code class="docutils literal notranslate"><span class="pre">BAM.BAI</span></code> format. BAM files are compressed binary forms of SAM files. SAM/BAM files are tab separated and show one alignment per line.
The information shown by the columns go as follows:
1. Read header
2. Bitwise FLAG
3. Name of reference sequence
4. Position of alignment (1-based)
5. MAPQ-score
6. CIGAR string
7. Name of mate read (shows * if information is not available)
8. Position of mate read (shows 0 if information is not available)
9. Length of alignment on the reference (shows 0 if information is not available)
10. Read sequence (shows * if information is not available)
11. Quality of read sequence (shows * if information is not available)</p>
<p>One of each is generated per sample.</p>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">3</span><span class="p">:</span><span class="mi">11409</span><span class="p">:</span><span class="mi">5471</span><span class="p">:</span><span class="mi">17963</span><span class="n">_AAGACACTG</span>     <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">14572</span>   <span class="mi">0</span>       <span class="mi">23</span><span class="n">M</span>     <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">CCACACAGTGCTGGTTCCGTCAC</span> <span class="n">EEEEEEEEEEEAEEEEEEEEEEE</span> <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">7</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">4</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">22</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
<span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">3</span><span class="p">:</span><span class="mi">11604</span><span class="p">:</span><span class="mi">9407</span><span class="p">:</span><span class="mi">1314</span><span class="n">_TCTGCCCAC</span>      <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">14747</span>   <span class="mi">0</span>       <span class="mi">36</span><span class="n">M</span>     <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">CGGCAGAGGAGGGATGGAGTCTGACACGCGGGCAAA</span>    <span class="n">EEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEE</span>    <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">5</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">4</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">35</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
<span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="mi">11201</span><span class="p">:</span><span class="mi">6526</span><span class="p">:</span><span class="mi">7382</span><span class="n">_TCCCCGACC</span>      <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">14847</span>   <span class="mi">0</span>       <span class="mi">40</span><span class="n">M</span>     <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">AGTGAGGGTGGTTGGTGGGAAACCCTGGTTCCCCCAGCCC</span>        <span class="n">EEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEE</span>        <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">6</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">3</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">39</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
<span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">11204</span><span class="p">:</span><span class="mi">3841</span><span class="p">:</span><span class="mi">14476</span><span class="n">_GCGATCCCG</span>     <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">14992</span>   <span class="mi">0</span>       <span class="mi">37</span><span class="n">M</span>     <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">GTTGAAGAGATCCGACATCAAGTGCCCACCTTGGCTC</span>   <span class="n">EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE</span>   <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">8</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">5</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">36</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
<span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="mi">11204</span><span class="p">:</span><span class="mi">16119</span><span class="p">:</span><span class="mi">17944</span><span class="n">_CACACCCCG</span>    <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">14992</span>   <span class="mi">0</span>       <span class="mi">37</span><span class="n">M</span>     <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">GTTGAAGAGATCCGACATCAAGTGCCCACCTTGGCTC</span>   <span class="n">EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE</span>   <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">8</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">5</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">36</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
<span class="n">NB501399</span><span class="p">:</span><span class="mi">129</span><span class="p">:</span><span class="n">HLW7VAFX2</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">21211</span><span class="p">:</span><span class="mi">6880</span><span class="p">:</span><span class="mi">4260</span><span class="n">_CCACAACTC</span>      <span class="mi">272</span>     <span class="mi">1</span>       <span class="mi">15923</span>   <span class="mi">0</span>       <span class="mi">1</span><span class="n">S25M659N10M</span>    <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">GACCACTTCCCTGGGAGCTCCCTGGACTGAAGGAGA</span>    <span class="n">AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE</span>    <span class="n">NH</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">7</span>  <span class="n">HI</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">3</span>  <span class="n">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">35</span> <span class="n">nM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>
</pre></div>
</div>
</div>
<div class="section" id="cross-link-sites-peak-called">
<span id="output-cross-link-sites-peak-called"></span><h2>Cross link sites peak called<a class="headerlink" href="#cross-link-sites-peak-called" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="raw-cross-link-sites">
<span id="output-cross-link-sites-raw"></span><h2>Raw cross link sites<a class="headerlink" href="#raw-cross-link-sites" title="Permalink to this heading"></a></h2>
<p>Directory that contains unmodified cross-link sites with all background noise remaining. Cross-link sites are provided in 3 different formats, which are separated in one directory each; <code class="docutils literal notranslate"><span class="pre">WIG</span></code>, <code class="docutils literal notranslate"><span class="pre">BIGWIG</span></code> and <code class="docutils literal notranslate"><span class="pre">BEDGRAPH</span></code>. Each format represents the same data.
Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
<div class="section" id="wig-wiggle">
<h3>WIG (Wiggle)<a class="headerlink" href="#wig-wiggle" title="Permalink to this heading"></a></h3>
<p>Format to represent genome-wide coverage that consists of one line per reference chromosome with the coverage listed below each in a tab separated manner.
Column 1 represents the position while column 2 represents the coverage at the current position.
For each sample 2 WIG files are generated - one representing cross-link events on the forward ond one on the reverse string which can be distinguished by the name. The amount of cross-link events on the reverse strand is displayed as negative.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">variableStep</span> <span class="n">chrom</span><span class="o">=</span><span class="n">reference_1</span> <span class="n">span</span><span class="o">=</span><span class="mi">1</span>
<span class="mi">2815</span>    <span class="mf">1.0</span>
<span class="mi">3726</span>    <span class="mf">1.0</span>
<span class="mi">3895</span>    <span class="mf">1.0</span>
<span class="mi">6201</span>    <span class="mf">1.0</span>
<span class="mi">6367</span>    <span class="mf">1.0</span>
<span class="n">variableStep</span> <span class="n">chrom</span><span class="o">=</span><span class="n">reference_2</span> <span class="n">span</span><span class="o">=</span><span class="mi">1</span>
<span class="mi">22</span>  <span class="mf">1.0</span>
<span class="mi">31</span>  <span class="mf">1.0</span>
<span class="mi">66</span>  <span class="mf">1.0</span>
<span class="mi">80</span>  <span class="mf">1.0</span>
</pre></div>
</div>
</div>
<div class="section" id="bigwig">
<h3>BIGWIG<a class="headerlink" href="#bigwig" title="Permalink to this heading"></a></h3>
<p>An extension of the previously mentioned WIG format. While WIG uses plain text BIGWIG uses a binary format to store the data, reducing the file size. Therefore, accessing the data requires specialized software such as the IGV.</p>
</div>
<div class="section" id="bedgraph">
<h3>BEDGRAPH<a class="headerlink" href="#bedgraph" title="Permalink to this heading"></a></h3>
<p>Similar format to WIG or BIGWIG. BEDGRAPH files consist of 4 columns:
1. The chromosome name
2. The start position of the described events
3. The end position of the described events (for PARANOiD this is the position of the actual cross-link event)
4. Coverage of currently described event (negative for reverse strand)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DQ375404</span><span class="mf">.1</span>  <span class="mi">2814</span>    <span class="mi">2815</span>    <span class="mi">1</span>
<span class="n">DQ375404</span><span class="mf">.1</span>  <span class="mi">3725</span>    <span class="mi">3726</span>    <span class="mi">1</span>
<span class="n">DQ375404</span><span class="mf">.1</span>  <span class="mi">3894</span>    <span class="mi">3895</span>    <span class="mi">1</span>
<span class="n">DQ375404</span><span class="mf">.1</span>  <span class="mi">6200</span>    <span class="mi">6201</span>    <span class="mi">1</span>
<span class="n">DQ375404</span><span class="mf">.1</span>  <span class="mi">6366</span>    <span class="mi">6367</span>    <span class="mi">1</span>
<span class="n">DQ380154</span><span class="mf">.1</span>  <span class="mi">21</span>      <span class="mi">22</span>      <span class="mi">1</span>
<span class="n">DQ380154</span><span class="mf">.1</span>  <span class="mi">30</span>      <span class="mi">31</span>      <span class="mi">1</span>
<span class="n">DQ380154</span><span class="mf">.1</span>  <span class="mi">65</span>      <span class="mi">66</span>      <span class="mi">1</span>
<span class="n">DQ380154</span><span class="mf">.1</span>  <span class="mi">79</span>      <span class="mi">80</span>      <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="visualization-with-igv">
<h3>Visualization with IGV<a class="headerlink" href="#visualization-with-igv" title="Permalink to this heading"></a></h3>
<p>All provided file types can be easily visualized via the <a class="reference external" href="https://software.broadinstitute.org/software/igv/">Integrative Genomics Viewer (IGV)</a>.
To do so first the reference sequences need to be loaded into IGV. This is done by clicking on the tab called <em>Genomes</em> - which is located on the top left corner - and then choose the origin of the reference genome.</p>
<p><a class="reference internal" href="_images/IGV-wig-overview.png"><img alt="Showing an overview of cross-link events via IGV" src="_images/IGV-wig-overview.png" style="width: 800px;" /></a></p>
<p>The reference track can be used to zoom in allowing users to see cross-link sites more detailed.</p>
<p><a class="reference internal" href="_images/IGV-wig-zoomed.png"><img alt="Zoom into cross-link events via IGV" src="_images/IGV-wig-zoomed.png" style="width: 800px;" /></a></p>
</div>
</div>
<div class="section" id="cross-link-sites-merged">
<span id="output-cross-link-sites-merged"></span><h2>Cross link sites merged<a class="headerlink" href="#cross-link-sites-merged" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="execution-metrics">
<span id="output-execution-metrics"></span><h2>Execution metrics<a class="headerlink" href="#execution-metrics" title="Permalink to this heading"></a></h2>
<p>Directory that contains general execution metrics of the workflow such as:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>container_information.txt</dt><dd><p>Container system used to execute the processes together with the containers that were used during the workflow</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>execution_information.txt</dt><dd><dl class="simple">
<dt>Contains information necessary to reproduce the results such as</dt><dd><ol class="loweralpha simple">
<li><p>Command used for the execution</p></li>
<li><p>Directory of PARANOiD</p></li>
<li><p>Config file used</p></li>
<li><p>Profiles used</p></li>
<li><p>Version od Nextflow and PARANOiD</p></li>
<li><p>Execution directory</p></li>
</ol>
</dd>
</dl>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>parameter_information.txt</dt><dd><p>Contains all parameters used</p>
</dd>
</dl>
</li>
</ol>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
<div class="section" id="igv-session">
<span id="output-igv-session"></span><h2>IGV-session<a class="headerlink" href="#igv-session" title="Permalink to this heading"></a></h2>
<p>An <code class="docutils literal notranslate"><span class="pre">XML</span></code> file that can be transferred directly to the IGV.
This can be done by clicking on the <em>data</em> tab on the top left and then on <em>Open Session</em>. A window will open in which you can navigate to the output directory of PARANOiD and choose <em>igv-session.xml</em>.
This will open a predefined IGV session that includes the reference, the cross-link sites of all samples (forward and reverse) and the alignment files of all samples.
If the option <a class="reference internal" href="PARAMETERS.html#merge-replicates"><span class="std std-ref">--merge_replicates</span></a> was chosen then only the merged cross-link sites are shown.
Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
<div class="section" id="peak-height-distribution">
<span id="output-peak-height-distribution"></span><h2>Peak height distribution<a class="headerlink" href="#peak-height-distribution" title="Permalink to this heading"></a></h2>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
<div class="section" id="reference">
<span id="output-reference"></span><h2>Reference<a class="headerlink" href="#reference" title="Permalink to this heading"></a></h2>
<p>The <a class="reference internal" href="INPUT.html#reference"><span class="std std-ref">reference sequence</span></a> provided as input.</p>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
<div class="section" id="statistics">
<span id="output-statistics"></span><h2>Statistics<a class="headerlink" href="#statistics" title="Permalink to this heading"></a></h2>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
<div class="section" id="strand-distribution">
<span id="output-strand-distribution"></span><h2>Strand distribution<a class="headerlink" href="#strand-distribution" title="Permalink to this heading"></a></h2>
<p>Is included in the <a class="reference internal" href="ANALYSES.html#basic-analysis"><span class="std std-ref">basic analysis</span></a>.</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ANALYSES.html" class="btn btn-neutral float-left" title="Included Analyses" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="CONTAINER.html" class="btn btn-neutral float-right" title="Container usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Patrick Barth.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>